src/encoding.ts

```diff
@@ -0,0 +1,333 @@
+import { HIGH_SURROGATE_MIN, HIGH_SURROGATE_MAX, LOW_SURROGATE_MIN, LOW_SURROGATE_MAX } from './string';
+
+type ByteBuffer = Uint8Array | Buffer | number[];
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+
+function validateDecodingArgs(data: any) {
+	return data == null ? false : true;
+}
+
+function validateUtf8EncodedCodePoint(codePoint: number): void {
+	if ((codePoint & 0xC0) !== 0x80) {
+		throw Error('Invalid continuation byte');
+	}
+}
+
+function decodeUtf8EncodedCodePoint(codePoint: number, validationRange: number[] = [ 0, Infinity ], checkSurrogate?: boolean): string {
+	if (codePoint < validationRange[0] || codePoint > validationRange[1]) {
+		throw Error('Invalid continuation byte');
+	}
+
+	if (checkSurrogate && codePoint >= 0xD800 && codePoint <= 0xDFFF) {
+		throw Error('Surrogate is not a scalar value');
+	}
+
+	let encoded = '';
+
+	if (codePoint > 0xFFFF) {
+		codePoint = codePoint - 0x010000;
+		encoded += String.fromCharCode(codePoint >>> 0x10 & 0x03FF | 0xD800);
+		codePoint = 0xDC00 | codePoint & 0x03FF;
+	}
+
+	encoded += String.fromCharCode(codePoint);
+
+	return encoded;
+}
+
+export interface Codec {
+	encode(data: string): number[];
+	decode(data: ByteBuffer): string;
+}
+
+/**
+ * Ascii codec that provides facilities for encoding a string
+ * into an Ascii encoded byte buffer as well as decoding an Ascii encoded byte buffer
+ * into a string.
+ */
+export const ascii: Codec = {
+	/**
+	 * Encode a string into an Ascii byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string = ''): number[] {
+		let buffer: number[] = [];
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			buffer[i] = data.charCodeAt(i);
+		}
+
+		return buffer;
+	},
+
+	/**
+	 * Decode an Ascii encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			decoded += String.fromCharCode(data[i]);
+		}
+
+		return decoded;
+	}
+}
+
+/**
+ * Utf8 codec that provides facilities for encoding a string
+ * into an Utf8 encoded byte buffer as well as decoding an Utf8 encoded byte buffer
+ * into a string.
+ * Inspired by the work of: https://github.com/mathiasbynens/utf8.js
+ */
+export const utf8: Codec = {
+	/**
+	 * Encode a string into an Utf8 byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string = ''): number[] {
+		let buffer: number[] = [];
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let encodedChar = data.charCodeAt(i);
+
+			/**
+			 * Surrogates
+			 * http://en.wikipedia.org/wiki/Universal_Character_Set_characters
+			 */
+			if (encodedChar > HIGH_SURROGATE_MIN && encodedChar < HIGH_SURROGATE_MAX) {
```

Comment:
The only problem with using `codePointAt` is that I need to keep track of what position I am at in the source string and that would be difficult with that utility function because it doesn't return any indicator of whether a low surrogate was provided or not. 

Also, according to this: http://people.mozilla.org/~jorendorff/es6-draft.html#sec-string.prototype.codepointat both of these ranges should both be exclusive which if thats correct then both this and `codePointAt` are wrong currently